packages/react-router/lib/hooks.tsx

```diff
@@ -432,7 +433,10 @@ export function useRoutes(
         })
       ),
     parentMatches,
-    dataRouterStateContext || undefined
+    // Only pass along the dataRouterStateContext when we're rendering from the
+    // RouterProvider layer.  If routes is different then we're rendering from
+    // a descendant <Routes> tree
+    dataRouterContext?.router.routes === routes ? dataRouterStateContext : null
```

Comment:
This is the root of the bug - `_renderMatches` was behaving as if in a data router _anywhere_ under the context, including in descendant `<Routes>`.  A potentially more explicit fix would be an optional prop to `useRoutes` but I was on the fence about touching the public API:

```
export function useRoutes(
  routes: RouteObject[],
  locationArg?: Partial<Location> | string,
  isDataRouterRender?: boolean,  // New arg passed true from <DataRoutes>
)
```